Update from 1541-celery_poc (#205)
workfloworchestrator/orchestrator-core: docs/architecture/application/scaling.md

# Scaling the Orchestrator

By default the Orchestrator is capable to handle a reasonable amount of workflows and tasks. For a larger and more
distributed workload we introduced the [Celery library](https://docs.celeryq.dev/en/stable/).

This document describes the two modes in which an Orchestrator instance can run and what you need to configure:

> 1. running tasks an nodes in a threadpool (default)

Reviewer: Hm, typo. This is meant to be 'workflows' but obviously I was thinking of something else when I types this.

2. running the Orchestrator with a number of workflow workers

## Running workflows or tasks within a threadpool

> This is the default configuration. There is no difference between workflows and tasks. They have the same priority.

Reviewer: Nice!

to different queues. Currently, there are two queues defined:

- workflows: starting or resuming workflows
- tasks: starting or resuming tasks

By default, [Redis](https://redis.io/) is used for the Celery Broker and Backend. See the next chapter 
about implementing on how to change this behaviour.

### Implementing the worker

The orchestrator-core needs to know what workflows a user has defined. This information is only available in the
actual application using the orchestrator-core. Here we will use an example as used withing SURF. The files is 
called tasks.py. First we define our own class derived from the Celery base class:

```python
class OrchestratorCelery(Celery):
    def on_init(self) -> None:
        from orchestrator import OrchestratorCore

        app = OrchestratorCore(base_settings=AppSettings())
        load_surf(app)  # This will load the workflows
```

The `load_surf` should be replaced by your own function that at least makes sure that all the workflows are imported
(which make sure that the are registered) so that the worker can recognize them. This is the minimum implementation

> you need, but you might want to add other initialization that is needed to execute workflows.

Reviewer 2: Just for consistency
```suggestion
you need, but you might want to add other initialisation that is needed to execute workflows.
```

> Next we instantiate celery using our own `OrchestratorCelery` class:

Reviewer 2: ```suggestion
Next we instantiate Celery using our own `OrchestratorCelery` class:
```

```python
broker = f"redis://{AppSettings().CACHE_HOST}:{AppSettings().CACHE_PORT}"
backend = f"rpc://{AppSettings().CACHE_HOST}:{AppSettings().CACHE_PORT}/0"

celery = OrchestratorCelery("proj", broker=broker, backend=backend, include=["orchestrator.services.tasks"])

celery.conf.update(result_expires=3600)
```

As you can see in the code above we are using Redis as broker. You can of course replace this by RabbitMQ or 
another broker of your choice. See the Celery documentation for more details.

> `"orchestrator.services.tasks" ` is the namespace in orchestartor-core where the Celery tasks can be found. At the

Reviewer 2: ```suggestion
`"orchestrator.services.tasks" ` is the namespace in orchestrator-core where the Celery tasks can be found. At the
```

moment 4 tasks are defined: 

1. `tasks.new_task`: start a new task (delivered on the Task queue)
2. `tasks.new_workflow`: start a new workflow (delivered on the Workflow queue)
3. `tasks.resume_task`: resume an existing task (delivered on the Task queue)
4. `tasks.resume_workflow`: resume an existing workflow (delivered on the Workflow queue)

> Finally, we initialize the orchestrator core:

Reviewer 2: ```suggestion
Finally, we initialise the orchestrator core:
```

```python
def init_celery() -> None:
    from orchestrator.services.tasks import initialise_celery

    initialise_celery(celery)

> # Needed if we load this as a Celery worker because in that case there is no 'main app'

Reviewer 2: ```suggestion
# Needed if we load this as a Celery worker because in that case, the application is not started with a user-specified top-level `__main__`  module

init_celery()
```

> The code above sets our local celery instance (which initializes the workflows) as the celery instance that is

Reviewer 2: ```suggestion
The code above sets our local Celery instance (which initialises the workflows) as the celery instance that is
```

> going to be used by the orchestrator-core. Without this code the orchestrator-core would be only aware of a limited

Reviewer 2: ```suggestion
going to be used by the orchestrator-core. Without this code, the orchestrator-core would be only aware of a limited
```

set of workflows that are part of orchestrator-core itself.

### Running locally

> If you want to test your application locally you have to start both the orchestrator-api and one ore more workers.

Reviewer 2: ```suggestion
If you want to test your application locally you have to start both the orchestrator-api and one or more workers.
```

For example:

Start the orchestrator api:

```bash

> EXECUTOR="celery" DATABASE_URI="postgresql://nwa:nwa@localhost/nwa-workflows" ./start.sh dev

Reviewer 2: This is probably too obscure for orchestrator-core users outside of SURF. Maybe just leave out the DATABASE_URI environment variable, as this is not mentioned in the README either.
```suggestion
EXECUTOR="celery" uvicorn --reload --host 127.0.0.1 --port 8080 main:app
```

```

Notice that we are setting `EXECUTOR` to `celery`. Without that variable the api resorts to the default threadpool.

Start a single worker that listens both on the `tasks` and `workflows` queue:

```bash

> DATABASE_URI="postgresql://nwa:nwa@localhost/nwa-workflows" celery -A surf.tasks  worker --loglevel=info -Q tasks,workflows

Reviewer 2: ```suggestion
celery -A surf.tasks  worker --loglevel=info -Q tasks,workflows
```

The queues that the orchestrator-api  uses to submit tasks is not configurable, but we need to specify the correct queue names here. Maybe add some explanation what `-Q` is for and that you can use it to make high/low-prio workers, by specifying different queues

Reviewer: I'll elaborate a bit on this

```

> Notice that `-A surf.tasks` indicates the namespace of the file that contains your 'celery' instance.

Reviewer 2: ```suggestion
Notice that `-A surf.tasks` indicates the module that contains your 'celery.Celery' instance.
```

```python
def init_celery() -> None:
    from orchestrator.services.tasks import initialise_celery

    initialise_celery(celery)

# Needed if we load this as a Celery worker because in that case there is no 'main app'
init_celery()
```

The code above sets our local celery instance (which initializes the workflows) as the celery instance that is
going to be used by the orchestrator-core. Without this code, the orchestrator-core would be only aware of a limited
set of workflows that are part of orchestrator-core itself.

### Running locally

If you want to test your application locally you have to start both the orchestrator-api and one or more workers.
For example:

Start the orchestrator api:

```bash
EXECUTOR="celery" uvicorn --reload --host 127.0.0.1 --port 8080 main:app
```

Notice that we are setting `EXECUTOR` to `celery`. Without that variable the api resorts to the default threadpool.

Start a single worker that listens both on the `tasks` and `workflows` queue:

```bash
celery -A surf.tasks  worker --loglevel=info -Q tasks,workflows
```

Notice that `-A surf.tasks` indicates the module that contains your 'celery.Celery' instance.

## Deployment

> Todo: explain on how to use orchestrator[api] en orchestrator[worker]

Reviewer 3: ?

Reviewer: Ah yes, can be removed now since after last week discussion we decided not to have separate packages since they don't have any benefit in actual footprint.
